lib/phoenix_live_view/html_engine.ex

```diff
@@ -1040,23 +1107,84 @@ defmodule Phoenix.LiveView.HTMLEngine do
     end
   end
 
-  defp store_component_call(module, component, attrs, file, line) do
+  defp store_component_call(module, component, slots, attrs, file, line, env) do
     if Module.open?(module) do
+      pruned_slots =
+        for {slot_name, slot_values} <- slots,
+            do: {slot_name, Enum.map(slot_values, &slot_call_value(&1, env))},
+            into: %{}
+
       pruned_attrs =
         for {attr, value, meta} <- attrs,
             is_binary(attr) and not String.starts_with?(attr, ":"),
-            do: {String.to_atom(attr), {meta[:line], meta[:column], component_call_value(value)}},
+            {kind, value} = component_call_value(value, env),
+            do: {String.to_atom(attr), {meta[:line], meta[:column], kind, value}},
             into: %{}
 
       root = List.keymember?(attrs, :root, 0)
-      call = %{component: component, attrs: pruned_attrs, file: file, line: line, root: root}
+
+      call = %{
+        component: component,
+        slots: pruned_slots,
+        attrs: pruned_attrs,
+        file: file,
+        line: line,
+        root: root
+      }
+
       Module.put_attribute(module, :__components_calls__, call)
     end
   end
 
-  defp component_call_value({:expr, _, _}), do: :expr
-  defp component_call_value({:string, string, _}), do: string
-  defp component_call_value(nil), do: nil
+  defp component_call_value({:string, value, _meta}, _env) do
+    {:lit, value}
+  end
+
+  defp component_call_value(nil, _env) do
+    {:lit, true}
+  end
+
+  defp component_call_value({:expr, value, %{line: line, column: column}}, env) do
+    case Code.string_to_quoted!(value, line: line, column: column, file: env.file) do
+      # lists
+      {ast, _, _} = value when is_list(ast) ->
+        {:list, value}
+
+      # structs
+      {:%, _, [{_, _, [mod]} | _]} ->
+        {:struct, lookup_alias(env, mod)}
```

Comment:
```suggestion
      # structs
      {:%, _, [{_, _, [mod]} | _]} ->
        {:struct, lookup_alias(env, mod)}
```

This can lead to bugs because we only have the environment from the beginning of the template. Unfortunately I don't think inferring the struct is possible. We should remove this clause and the struct checks from the compiler. Sorry. :(

Reply:
> This can lead to bugs because we only have the environment from the beginning of the template.

Could you elaborate on this? Is this a limitation of colocated template files, or how HEEx and EEx work in general?

Reply:
It is a limitation of engines. It could be worked around, but it would require too much to work. Basically the idea is that if you have this code:

```elixir
  ~H"""
  <%= macro_that_defines_alias() %>
  <.component bar={%Foo{}} />
  """
```

The environment that we have represents the code environment before `~H"""`. Any import, alias, var, etc that is defined within the template won't be reflected in the env, which would be the source of bugs I mentioned above. So we can expand aliases to the wrong thing.

We could fix by also expanding and updating the environment as we move forward, but that's too much work and there would likely be corner cases too. So my suggestion is to skip the struct check for now (especially because i think most structs will not be literals in the template anyway).

Reply:
Oh I understand now, thanks for the explanation 😁 I'll remove the struct literal checks from this PR and update the documentation for the `attr` macro.